src/dbnode/storage/bootstrap/bootstrapper/commitlog/source.go

```diff
@@ -782,6 +782,14 @@ func (s *commitLogSource) newReadCommitLogPred(
 	// that has a snapshot more recent than the global minimum. If we use an array for fast-access this could
 	// be a small win in terms of memory utilization.
 	return func(f commitlog.File) bool {
+		if f.Error != nil {
```

Comment:
We can't skip corrupt commit log files like they didn't exist w/o risking data loss. It'd be much safe to see if any corrupt commit logs exist and if so fall back to the peers bootstrapper (in the case there are peers), and if there are no peers allow users to opt in to skip corrupt commit log files but don't do it by default.

Reply:
I'm ok with making this opt-outable for users who really care about data integrity, but I think by default it should skip them. If you can't even read the info part of the commitlog, then (barring true corruption of the filesystem unrelated to M3DB shutdown) your commitlog isn't missing any data anyways other than the standard amount you would be missing due to the writebehind strategy. I don't think this is a situation where we need to involve operators in the general case.

Reply:
by putting the commitlog bootstrapper before peers bootstrapper you're basically accepting losing the last few writes that occurred right before shutdown due to the writebehind strategy (as well as any writes that were issued while you were down) anyways.

In my mind we should emit a log, skip over bad commitlog files, and then eventually address this issue with the corruption APIs / a background repair process.

Reply:
The other reason I think it sucks to push this burden down to the operators is that basically by the time this happen and you realize whats going on, there are only two options:

1. Disable commitlog bootstrapper and replace it with peers bootstrapper, which is fine, but only viable in non-catastrophic scenarios where all the other nodes are up.

2. Go and delete the bad commitlog file and then restart with the commitlog bootstrapper, at which point you've probably written out enough commit log files that what should have been 6->8 minutes of downtime turns into an hour-long ordeal.

Reply:
For the peers case: how about just returning returning all requested shard ranges as unfullfilled in the commit log bootstrapper (won't need any interface changes from the way things are now) when the commit log runs into an error and that way we'll fall back to peers automatically. 

For the single node case: we can have a new RunOption to control the behaviour of whether to skip or not and the bootstrap process can be configured from the top level to specify what to pass it. That way the single node config we put out can default to come up but we can put a `WARNING` next to it and in docs.

Reply:
Spoke offline and settled on:

```
1. Add static configuration for whether or not commitlog bootstrapping corruption should return unfulfilled or not for the whole thing (default to return unfulfilled)
2. Always read all the commit log files we intended to read anyways, regardless of that config.
3. Add exception in commitlog bootstrap for single node deployment which will detect that this is the only node in the placement and default to returning fulfilled because there is nothing else you could possibly do to recover data
4. Don't return errors from the commitlog bootstrapper for corrupt files
5. Emit metrics / logs for all encountered corrupt files.
```

This gives us a good mix between maintaining data integrity as much as we can, but also not imposing undue operational burden.

In single node failures, the issue will just resolve itself.
In a single node failure within a multi-node cluster, the issue will just resolve itself.
In a multi-node failure within a multi-node cluster, *some* nodes with corrupt commitlog files may get stuck in the peer bootstrapping phase, but they can be "released" with a K.V change to the peer bootstrapper consistency level and come online without requiring an additional costly restart.